pipelines/google_sheets/google_sheets.py

```diff
@@ -0,0 +1,248 @@
+# Contains the main pipeline functions
+
+from typing import Any, Iterator, cast
+import dlt
+from dlt.common.configuration.specs import GcpClientCredentialsWithDefault
+from dlt.common.typing import DictStrAny, TDataItem
+from dlt.common.exceptions import MissingDependencyException
+import logging
+from pipelines.google_sheets.data_validator import serial_date_to_datetime, get_spreadsheet_id, get_first_rows, get_data_type
+
+try:
+    from apiclient.discovery import build, Resource
+except ImportError:
+    raise MissingDependencyException("Google Sheets Source", ["google1", "google2"])
+
+# TODO: consider using https://github.com/burnash/gspread for spreadsheet discovery
+# TODO: add exceptions to the code in case HTTP requests fail
+
+
+def get_metadata(spreadsheet_id: str, service: Resource, ranges: list[str]) -> Iterator[DictStrAny]:
+    """
+    Gets the metadata for the first 2 rows of every range specified. The first row is deduced as the header and the 2nd row specifies the format the rest of the data should follow
+    @:param: spreadsheet_id - the id of the spreadsheet
+    @:param: service - Resource object used by google-api-python-client to make api calls
+    @:param: ranges - List of ranges to get data from. If left empty, every sheet inside the spreadsheet will be included instead. named ranges not supported
+    @:return: ranges_data - A dict where all the range names are the key. The values for each key are the corresponding sheet metadata: sheet_name, headers, values
+    """
+
+    # process metadata ranges so only the first 2 rows are appended
+    ranges_data = {}
```

Comment:
this is extremely complicated and hard to test. should we just call the API many times for each range we want to get metadata? it won't impact the performance significantly

imo this is premature optimization